node/activation_service.go

```diff
@@ -0,0 +1,712 @@
+		}
+		poetClients = append(poetClients, client)
+	}
+	app.poetClients = poetClients
+
+	nipostBuilder, err := activation.NewNIPostBuilder(
+		app.localDB,
+		grpcPostService.(*grpcserver.PostService),
+		nipostLogger,
+		app.Config.POET,
+		app.clock,
+		app.validator,
+		activation.NipostbuilderWithPostStates(postStates),
+		activation.NipostbuilderWithIdentityStates(app.idStates),
+		activation.WithPoetServices(poetClients...),
+	)
+	if err != nil {
+		return fmt.Errorf("create nipost builder: %w", err)
+	}
+
+	builderConfig := activation.Config{
+		GoldenATXID:      goldenATXID,
+		RegossipInterval: app.Config.RegossipAtxInterval,
+	}
+
+	var (
+		atxBuilderLog = app.addLogger(ATXBuilderLogger, lg).Zap()
+		syncer        activation.Syncer
+	)
+	atxService := nodeServiceClient
+	atxPublisher := nodeServiceClient
+	syncer = alwaysSyncedSyncer{}
+
+	atxBuilder := activation.NewBuilder(
+		builderConfig,
+		app.localDB,
+		atxService,
+		atxPublisher,
+		app.validator,
+		nipostBuilder,
+		app.clock,
+		syncer,
+		atxBuilderLog,
+		activation.WithContext(ctx),
+		activation.WithPoetConfig(app.Config.POET),
+		// TODO(dshulyak) makes no sense. how we ended using it?
+		activation.WithPoetRetryInterval(app.Config.HARE3.PreroundDelay),
+		activation.WithPostStates(postStates),
+		activation.WithIdentityStates(app.idStates),
+		activation.WithPoets(poetClients...),
+		activation.BuilderAtxVersions(app.Config.AtxVersions),
+	)
+
+	if len(app.signers) > 1 || app.signers[0].Name() != supervisedIDKeyFileName {
+		// in a remote setup we register eagerly so the atxBuilder can warn about missing connections asap.
+		// Any setup with more than one signer is considered a remote setup. If there is only one signer it
+		// is considered a remote setup if the key for the signer has not been sourced from `supervisedIDKeyFileName`.
+		//
+		// In a supervised setup the postSetupManager will register at the atxBuilder when
+		// it finished initializing, to avoid warning about a missing connection when the supervised post
+		// service isn't ready yet.
+		for _, sig := range app.signers {
+			atxBuilder.Register(sig)
+		}
+	}
+	app.postSupervisor = activation.NewPostSupervisor(
+		app.log.Zap(),
+		app.Config.POST,
+		app.Config.SMESHING.ProvingOpts,
+		postSetupMgr,
+		atxBuilder,
+	)
+	if err != nil {
+		return fmt.Errorf("init post service: %w", err)
+	}
+
+	app.mesh = msh
+	app.syncer = newSyncer
+	app.atxBuilder = atxBuilder
+	app.atxHandler = atxHandler
+	app.poetDb = poetDb
+	app.fetcher = fetcher
+	app.beaconProtocol = beaconProtocol
+
+	return nil
+}
+
+func (app *App) startActivationServiceServices(ctx context.Context) error {
+	if app.fetcher != nil {
+		if err := app.fetcher.Start(); err != nil {
+			return fmt.Errorf("start fetcher: %w", err)
+		}
+	}
+	if app.syncer != nil {
+		app.syncer.Start()
+	}
+
+	if app.beaconProtocol != nil {
+		app.beaconProtocol.Start(ctx)
+	}
```

Comment:
How about inverting the check and _require_ that it's not nil? It would catch problems with initialization soon instead of some part just silently not working (not started).
```suggestion
	if app.beaconProtocol != nil {
		return errors.New("beacon service cannot be nil")
	}
	app.beaconProtocol.Start(ctx)
```